proposals/1859_github-auth/README.md

```diff
@@ -0,0 +1,500 @@
+
+There are other differences which would need to be documented for the user who is configuring the GitHub App or
+OAuth App, but those differences would not change the Pinniped code. The user would need to know how each of these two
+app types are installed or approved by a GitHub org, because without taking the appropriate steps for the org,
+the `/user/orgs` GitHub API will not return any results related to that org. This would prevent users of that org from
+being able to authenticate, because Pinniped would not be able to discover that they are members of that org.
+
+During initial login (authorization) the Supervisor can get an access token from GitHub and use it to learn the
+identity, org memberships, and team memberships of that user by calling the GitHub API. It can then store that
+access token into the Supervisor's session storage. During the Supervisor session refresh flow, it can make the same
+calls to the GitHub API to ensure that the access token is still valid, the user's identity has not changed,
+the user's org membership has not changed, and to update the users groups by getting their current team memberships.
+
+The three GitHub API requests during login, and the three more GitHub API requests during each refresh, will all count
+against that GitHub user's hourly API request limits. Note that refreshes only happen when users are actively
+interacting with Kubernetes clusters during their session.
+For one session started using browser-based auth, and used actively throughout an hour, this would be a maximum
+of approximately 36 GitHub API requests per hour against the user's 5,000 requests per hour limit.
+Additionally, each new concurrent Supervisor session started
+by that same user will cause the same API calls again (new sessions are started when the user's home directory is not
+shared between them, e.g. because the user is authenticating from multiple computers at the same time).
+
+#### CLI-based Authentication using GitHubIdentityProvider
+
+Non-interactive authentication at the CLI (no web browser) is desirable for scripting and CI/CD use cases.
+Some enterprises prefer to create "bot" users in their regular identity providers and use Pinniped for auth,
+rather than using Kubernetes' other features for authentication of non-human actors. OIDCIdentityProviders,
+LDAPIdentityProviders, and ActiveDirectoryIdentityProviders already support CLI-based auth, so it would be
+preferable for GitHubIdentityProvider to also support it, if possible.
+
+Note that non-interactive auth is only allowed on the `pinniped-cli` client, not on any configured
+third-party OIDCClient.
+
+Since we will use the GitHub API to discover a user's identity, we could use a PAT provided by the user
+at the CLI to authenticate that user. To keep the experience (and code) similar to OIDC and LDAP/AD identity providers,
+the CLI could interactively prompt for their username and password, and the user could paste the PAT as their password.
```

Comment:
Is it the intent of Github to provide a `username` along with a PAT, or is the PAT intended to  be standalone?

Reply:
Ah, good catch. I added some text to propose how to handle this.